src/ForwardedHttpMiddleware.cs

```diff
@@ -1,27 +1,507 @@
+                {
+                    if (!string.IsNullOrEmpty(set.Host) && TryValidateHost(set.Host)
+                        && (_allowAllHosts || HostString.MatchesAny(set.Host, _allowedHosts)))
+                    {
+                        applyChanges = true;
+                        currentValues.Host = set.Host;
+                    }
+                }
+            }
+
+            if (applyChanges)
+            {
+                var feature = new ForwardedHttpFeature();
+
+                if (checkFor && !string.IsNullOrEmpty(currentValues.RemoteIpAndPortText))
+                {
+                    if (connection.RemoteIpAddress != null)
+                    {
+                        // Save the original
+                        // WHERE!? In a header? in another feature?
+                    }
+
+                    if (currentValues.ForType == NodeType.IP)
+                    {
+                        connection.RemoteIpAddress = currentValues.RemoteIpAndPort.Address;
+                        connection.RemotePort = currentValues.RemoteIpAndPort.Port;
+
+                    }
+                    else
+                    {
+                        connection.RemoteIpAddress = null;
+                        connection.RemotePort = 0;
+                    }
+                    
+                    feature.ForType = currentValues.ForType;
+                    feature.For = currentValues.RemoteIpAndPortText;
+                }
+
+                if (checkBy && currentValues.LocalIpAndPort != null)
+                {
+                    if (connection.LocalIpAddress != null)
+                    {
+                        // Save the original??
+                    }
+
+                    connection.LocalIpAddress = currentValues.LocalIpAndPort.Address;
+                    connection.LocalPort = currentValues.LocalIpAndPort.Port;
+                }
+
+                if (checkProto && currentValues.Scheme != null)
+                {
+                    // Save the original??
+
+                    request.Scheme = currentValues.Scheme;
+                }
+
+                if (checkHost && currentValues.Host != null)
+                {
+                    // Save the original??
+
+                    request.Host = HostString.FromUriComponent(currentValues.Host);
+                }
+
+                context.Features.Set<IForwardedHttpFeature>(feature);
+
+                if (forwardedHeader.ForwardedValues.Count > entriesConsumed)
+                {
+                    // Truncate the consumed header values
+                    forwardedHeader.ForwardedValues = forwardedHeader.ForwardedValues.Take(forwardedHeader.ForwardedValues.Count - entriesConsumed).ToList();
+                }
+                else
+                {
+                    // All values were consumed
+                    forwardedHeader.ForwardedValues.Clear();
+                }
+
+                request.Headers.Remove("Forwarded");
```

Comment:
Make `Forwarded` a constant.

Reply:
Working with YARP I realized that removing the processed headers was probably a design mistake, it breaks the append-and-forward design intended for these headers. We should re-think how this would work end-to-end in a proxy like YARP. https://github.com/microsoft/reverse-proxy/issues/642